### DISCUSSION TITLE
Computed properties as command's aggregate uuid

### DISCUSSION THREAD

**Discussion by attac-t** [spatie/laravel-event-sourcing#487] [Q&A]
I was trying out an idea which makes use of `spatie/laravel-data` and this event-sourcing package where I was playing out with having the command class extending the `Data` class and  act as a `request` in itself.

Let's take as an example a command `CreatePerson`. Usually from the controller I generate a `ulid` string and assign it as the `personId` within the command. Instead, this approach tries to leverage the `Spatie\LaravelData\Attributes\Computed` attribute to generate and assign the `ulid` within the command itself.

The problem arises when the same `computed` property is also assigned the `AggregateUuid` attribute as it throws a `Spatie\ EventSourcing \ AggregateRoots \ Exceptions \ MissingAggregateUuid` exception.

I'm wondering if doing it this way is against how these two packages are meant to work together which is why I'm asking this here. Should this be acknowledged, I'm happy to contribute to a fix.

```php
#[HandledBy(Person::class)]
class CreatePerson extends Data
{
    #[AggregateUuid, Computed]
    public readonly ?string $person_id;

    public function __construct(
        public readonly string $first_name,
        public readonly string $last_name,
    ) {
        $this->person_id = (string) Str::ulid();
    }
}
```

```php
class PersonController
{
    public function __construct(
        protected readonly CommandBusContract $commandBus
    ) {
    }

    public function store(CreatePerson $command)
    {
        DB::transaction(fn() => $this->commandBus->dispatch($command));

        return response()->noContent();
    }
}
```

**Comment by inmanturbo** [comment]
It seems I've done something similar but not read-only. Is it not just that aggregate uuid is expected in the constructor? 

I've had some success using an aggregate uuid set to null in the constructor signature than simply creating the uuid within the constructor, but then  of course it's not quite as nice.

**Comment by inmanturbo** [comment]
I use LaravelData for both events and commands (using the traits instead of extending Data) and I very much like your approach above.

**Reply by attac-t** [reply]
So is it something that we should look into fixing? Given that you're already a contributor to this package, maybe you have some insight to answer this question.

**Reply by inmanturbo** [reply]
The command/bus implementation in this package hasn't seen much development in awhile. I am under the impression that the maintainers don't use the pattern much in on their own projects, given some of the issues with concurrency that have yet to be resolved, for which I believe they are waiting to address in a major point release (if ever).

That said, it may be worth a PR to take a stab at it. Spatie has been known to favor cleaner and stricter patterns. The readonly/computed uuid is bound to have at least some appeal.
